<!DOCTYPE html>

<html>
<head>
<title>Synthetic Data</title>
<meta charset="utf-8"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<link href="../static/favicon.ico" rel="icon" type="image/x-icon"/>
<link href="../static/page.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<nav>
<div class="row">
<div class="col-2 left">
<a href="../">Home</a>
</div>
<div class="col-10 right">
<a href="../license.html">License</a>
	  ·
	  <a href="../bibliography.html">Bibliography</a>
	  ·
	  <a href="../glossary.html">Glossary</a>
</div>
</div>
</nav>
<main>
<h1>Synthetic Data</h1>
<h2>The Problem</h2>
<ul>
<li>How do you test a data analysis pipeline?<ul>
<li>Unit tests of the kind used for invasion percolation aren't necessarily appropriate</li>
<li>Problem isn't the control flow of the code but how it interacts with data</li>
</ul>
</li>
<li><a href="../bibliography.html#Alexander2023">Alexander2023</a>: build a <a href="../glossary.html#synth_data">synthetic data generator</a><ul>
<li>Check the behavior of the pipeline by pushing data with known properties through it</li>
<li>If it finds something statistically significant in white noise, rethink the pipeline</li>
</ul>
</li>
<li><a href="https://faker.readthedocs.io/">Faker</a> has tools for generating a lot of useful data (we will use it in future chapters)</li>
<li>But it can't generate genomes, so we'll build our own</li>
</ul>
<p>FIXME: replace pandas with polars</p>
<hr/>
<h2>Synthesizing Genomes</h2>
<ul>
<li>Create some short snail genomes with <a href="../glossary.html#snp">single nucleotide polymorphisms</a></li>
<li>Output will have:<ul>
<li>Length: for now, all sequences the same length)</li>
<li>Reference sequence: the unmutated original</li>
<li>Locations: where mutations can occur</li>
<li>Susceptible location and base: which of those locations is the key to variance,
    and what mutated base causes the visible mutation</li>
<li>Individuals: list of generated genomes</li>
</ul>
</li>
<li>Since this has multiple fields, store as JSON</li>
</ul>
<hr/>
<h2>Main Driver</h2>
<pre class="language-py"><code class="language-py" data-file="synthesize_genomes.py:main">def main():
    """Main driver."""
    args = parse_args()
    random.seed(args.seed)
    genomes = random_genomes(
        args.length,
        args.num_genomes,
        args.num_snp,
        args.prob_other,
    )
    add_susceptibility(genomes)
    save(args.outfile, genomes)
</code></pre>
<hr/>
<h2>Random Genomes</h2>
<ul>
<li>Create a random base sequence as the reference genome</li>
<li>Create duplicates for individuals</li>
<li>Determine where SNPs can occur</li>
<li>Introduce significant mutations</li>
<li>Introduce other random mutations</li>
<li>Sort for reproducibility <a href="../bibliography.html#Taschuk2017">Taschuk2017</a></li>
</ul>
<hr/>
<h2>Random Genomes</h2>
<pre class="language-py"><code class="language-py" data-file="synthesize_genomes.py:random_genomes">def random_genomes(length, num_genomes, num_snp, prob_other):
    """Generate a set of genomes with specified number of point mutations."""
    assert 0 &lt;= num_snp &lt;= length

    # Reference genomes and specific genomes to modify.
    reference = random_bases(length)
    individuals = [reference] * num_genomes

    # Locations for SNPs.
    locations = random.sample(list(range(length)), num_snp)

    # Introduce significant mutations.
    for loc in locations:
        candidates = _other_bases(reference, loc)
        bases = [reference[loc]] + random.sample(candidates, k=len(candidates))
        individuals = [_mutate_snps(reference, ind, loc, bases) for ind in individuals]

    # Introduce other random mutations.
    other_locations = list(set(range(length)) - set(locations))
    individuals = [
        _mutate_other(ind, prob_other, other_locations) for ind in individuals
    ]

    # Return structure.
    individuals.sort()
    locations.sort()
    return GenePool(
        length=length, reference=reference, individuals=individuals, locations=locations
    )
</code></pre>
<hr/>
<h2>Random Bases and Mutations</h2>
<pre class="language-py"><code class="language-py" data-file="synthesize_genomes.py:random_bases">def random_bases(length):
    """Generate a random sequence of bases of the specified length."""
    assert 0 &lt; length
    return "".join(random.choices(DNA, k=length))
</code></pre>
<pre class="language-py"><code class="language-py" data-file="synthesize_genomes.py:mutate_snps">def _mutate_snps(reference, genome, loc, bases):
    """Introduce single nucleotide polymorphisms at the specified location."""
    choice = _choose_one(bases, SNP_PROBS)
    return genome[:loc] + choice + genome[loc + 1 :]
</code></pre>
<pre class="language-py"><code class="language-py" data-file="synthesize_genomes.py:mutate_other">def _mutate_other(genome, prob, locations):
    """Introduce other mutations at specified locations."""
    if random.random() &gt; prob:
        return genome
    loc = random.sample(locations, k=1)[0]
    base = random.choice(_other_bases(genome, loc))
    genome = genome[:loc] + base + genome[loc + 1 :]
    return genome
</code></pre>
<hr/>
<h2>Synthesizing Samples</h2>
<ul>
<li>Our model is that snail size depends on:<ul>
<li>Presence of significant mutation</li>
<li>Distance from epicenter of pollution</li>
</ul>
</li>
<li>Building this forces us to be explicit about our model<ul>
<li>Variance depends on a single nucleotide</li>
<li>And on linear distance from a center point</li>
</ul>
</li>
<li>So:<ul>
<li>Load the individuals</li>
<li>Put each individual somewhere near the spill site</li>
<li>Generate a random reading for its size that depends on two factors</li>
</ul>
</li>
</ul>
<hr/>
<h2>Once More With Feeling</h2>
<ul>
<li>Main driver should be starting to look familiar</li>
</ul>
<pre class="language-py"><code class="language-py" data-file="synthesize_samples.py:main">def main():
    """Main driver."""
    args = parse_args()
    random.seed(args.seed)
    genomes = json.loads(Path(args.genomes).read_text())
    geo_params = get_geo_params(args)
    samples = generate_samples(args, genomes, geo_params)
    save(args, samples)
</code></pre>
<hr/>
<h2>Geography</h2>
<ul>
<li>Get geographic parameters from CSV files</li>
<li>Need to join tables to get longitude, latitude, and nominal pollution radius</li>
</ul>
<pre class="language-csv"><code class="language-csv" data-file="sites.csv">site,lon,lat
COW,-124.04519,48.82172
YOU,-124.197,48.87251
HMB,-124.17555,48.81673
GBY,-124.4593,48.9209
</code></pre>
<pre class="language-csv"><code class="language-csv" data-file="surveys.csv">label,site,date,num,peak,relative_sd,radius
1748,COW,2023-04-27,23,100.0,0.1,0.1
1749,COW,2023-04-28,11,100.0,0.1,0.1
1755,COW,2023-05-13,15,101.0,0.11,0.1
1781,YOU,2023-05-01,12,90.0,0.15,0.15
1790,HMB,2023-05-02,19,107.0,0.22,0.11
1803,GBY,2023-05-08,8,95.0,0.1,0.14
</code></pre>
<hr/>
<h2>Geography</h2>
<pre class="language-py"><code class="language-py" data-file="synthesize_samples.py:get_geo_params">def get_geo_params(args):
    """Get geographic parameters."""
    sites = pd.read_csv(Path(args.paramsdir, "sites.csv"))
    surveys = pd.read_csv(Path(args.paramsdir, "surveys.csv"))
    combined = sites.merge(surveys, how="inner", on="site")
    filtered = combined[combined["site"] == args.site].iloc[0]
    return {
        "lon": filtered["lon"],
        "lat": filtered["lat"],
        "radius": filtered["radius"],
    }
</code></pre>
<hr/>
<h2>Generate Locations and Sizes</h2>
<ul>
<li>Generate location and snail size based on genetics and distance</li>
</ul>
<pre class="language-py"><code class="language-py" data-file="synthesize_samples.py:generate_samples">def generate_samples(args, genomes, geo_params):
    """Generate snail samples."""
    samples = []
    for sequence in genomes["individuals"]:
        point, distance = random_geo_point(**geo_params)
        if sequence[genomes["susceptible_loc"]] == genomes["susceptible_base"]:
            limit = args.mutant
        else:
            limit = args.normal
        scale = limit * distance / geo_params["radius"]
        reading = random.uniform(
            MIN_SNAIL_SIZE, MIN_SNAIL_SIZE + MAX_SNAIL_SIZE * scale
        )
        samples.append((point.longitude, point.latitude, sequence, reading))

    df = pd.DataFrame(samples, columns=("lon", "lat", "sequence", "reading"))
    df["lon"] = df["lon"].round(LON_LAT_PRECISION)
    df["lat"] = df["lat"].round(LON_LAT_PRECISION)
    df["reading"] = df["reading"].round(SNAIL_PRECISION)

    return df
</code></pre>
<hr/>
<h2>Magic Numbers</h2>
<ul>
<li>Keep the magic numbers at the top of the file<ul>
<li>If we wanted to vary these, would store them as JSON or YAML and load</li>
<li>In particular, would do this if we were sweeping parameters as in <a href="../06_scale/index.html">the earlier chapter</a></li>
</ul>
</li>
</ul>
<pre class="language-py"><code class="language-py" data-file="synthesize_samples.py:parameters">CIRCLE = 360.0
LON_LAT_PRECISION = 5
READING_PRECISION = 1
MIN_SNAIL_SIZE = 0.5
MAX_SNAIL_SIZE = 5.0
SNAIL_PRECISION = 1
</code></pre>
<hr/>
<h2>Analysis</h2>
<ul>
<li>Finally ready to write our analysis</li>
<li>Read the CSV data with locations, genomes, and readings</li>
<li>Find all candidate locations where sequences don't agree</li>
<li>Create a scatter plot by location and base<ul>
<li>One showing all data</li>
<li>One showing only locations where there are variations</li>
</ul>
</li>
</ul>
<hr/>
<h2>All Our Snails</h2>
<figure id="mutate_all_scatter">
<img alt="scatterplot of all readings at all locations" src="all_data_scatter.svg"/>
<figcaption>Figure 1: Reading as a function of location and base (all)</figcaption>
</figure>
<hr/>
<h2>Snails With Mutations</h2>
<figure id="mutate_slimmed_scatter">
<img alt="scatterplot of readings for snails with mutations" src="slimmed_data_scatter.svg"/>
<figcaption>Figure 2: Reading as a function of location and base (mutants)</figcaption>
</figure>
<hr/>
<h2>Rank Order</h2>
<figure id="mutate_slimmed_sorted">
<img alt="rank plot of readings at mutation locations" src="slimmed_data_sorted.svg"/>
<figcaption>Figure 3: Reading as a function of location (rank order)</figcaption>
</figure>
<hr/>
<h2>Conclusions</h2>
<ul>
<li>Clearly a winner in this plot…</li>
<li>…but not nearly as clear in scatter plot…</li>
<li>…and the winner has only twice the reading of the next-highest (random) value</li>
<li>Statistics could tell us if this is what we expect,
    but this isn't a statistics lesson</li>
</ul>
<hr/>
<h2>Exercises</h2>
<p>FIXME: create exercises for data analysis chapter</p>
</main>
<footer>
<a href="../">The Webonomicon</a>
      copyright © 2024
      <a href="../contributing.html#contributors">the authors</a>
</footer>
</body>
</html>